File: src/main/scala/bad/robot/radiate/ui/CompositeLayerUI.scala
package bad.robot.radiate.ui

import java.awt.Graphics
import javax.swing._
import javax.swing.plaf.LayerUI

object CompositeLayerUI {
  def apply[C <: JComponent](layers: List[LayerUI[C]]): LayerUI[C] = {
    new CompositeLayerUI[C](layers)
  }
}

class CompositeLayerUI[C <: JComponent](layers: List[LayerUI[C]]) extends LayerUI[C] {

  override def installUI(component: JComponent) {
    layers.foreach(layer => layer.installUI(component))
    super.installUI(component)
  }

  override def paint(graphics: Graphics, component: JComponent) {
    super.paint(graphics, component)
    layers.foreach(layer => layer.paint(graphics, component))
  }

  override def uninstallUI(component: JComponent) {
    layers.foreach(layer => layer.uninstallUI(component))
    super.uninstallUI(component)
  }
}

File: src/main/java/bad/robot/radiate/ui/CompositeLayerUI.java
package bad.robot.radiate.ui;

import javax.swing.*;
import javax.swing.plaf.LayerUI;
import java.awt.*;
import java.util.Arrays;
import java.util.List;

@Deprecated
public class CompositeLayerUI<C extends JComponent> extends LayerUI<C> {
    
    private final List<LayerUI<C>> layers;

    public static <C extends JComponent> LayerUI<C> layers(LayerUI<C>... layers) {
        return new CompositeLayerUI<>(layers);
    }

    public CompositeLayerUI(LayerUI<C>... layers) {
        this.layers = Arrays.asList(layers);
    }

    @Override
    public void installUI(JComponent component) {
        layers.forEach(layer -> layer.installUI(component));
        super.installUI(component);
    }

    @Override
    public void paint(Graphics graphics, JComponent component) {
        super.paint(graphics, component);
        layers.forEach(layer -> layer.paint(graphics, component));
    }

    @Override
    public void uninstallUI(JComponent component) {
        layers.forEach(layer -> layer.uninstallUI(component));
        super.uninstallUI(component);
    }
}
